### PR TITLE
Provide consistent interface for CloudExport API

### PR DESCRIPTION
**Pull request by danielfurman**
Issue: KNTK-377

### CODE REVIEW

**Review by jdziurzynski — APPROVED** [review]

File: examples/cloud_export_example_test.go [hunk 1]
```diff
@@ -1,85 +1,253 @@
...
+	ce.Type = models.CloudExportTypeKentikManaged
+	ce.Enabled = pointer.ToBool(true)
+	ce.Description = "Dummy AWS description"
+	ce.APIRoot = "https://dummy.api.kentik.com"
+	ce.FlowDestination = "https://dummy.flow.kentik.com"
+	ce.AWSProperties.IAMRoleARN = "dummy-iam-role-arn"
+	ce.AWSProperties.Region = "dummy-region"
+	ce.AWSProperties.DeleteAfterRead = pointer.ToBool(true)
+	ce.AWSProperties.MultipleBuckets = pointer.ToBool(true)
+	ce.BGP = &models.BGPProperties{
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go:75]: isn't it either AWS or BGP? I might be wrong but I thing they are alternatives and the one user is interested in is specified in CloudProvider field

Comment (mateuszmidor) [reply examples/cloud_export_example_test.go:75]: there should probably be a separate "demonstrateCloudExportAPIWithBGP"

File: examples/cloud_export_example_test.go [hunk 2]
```diff
@@ -1,85 +1,253 @@
...
-	fmt.Println("Invoking client.CloudExportAdmin.ListCloudExport")
-	getAllResp, err := client.CloudExportAdmin.ListCloudExport(ctx, &cloudexportpb.ListCloudExportRequest{})
+	fmt.Println("Getting all cloud exports")
+	getAllResp, err := client.CloudExports.GetAll(ctx)
+	if err != nil {
+		return fmt.Errorf("client.CloudExports.GetAll: %w", err)
+	}
+
+	fmt.Println("Invalid cloud exports count:", getAllResp.InvalidCloudExportsCount)
+	fmt.Printf("Listed cloud exports: %+v\n", getAllResp.CloudExports)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: For complex structures, PrettyPrint gives more useful output - including aggregated structures instead of just pointers

File: examples/cloud_export_example_test.go [hunk 3]
```diff
@@ -1,85 +1,253 @@
...
+
+	fmt.Println("Creating AWS cloud export")
+	ce := models.NewAWSCloudExport(models.CloudExportAWSRequiredFields{
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go:59]: I like the cloud-dedicated constructors

File: examples/cloud_export_example_test.go [hunk 4]
```diff
@@ -1,85 +1,253 @@
...
+		Name:          "example-aws-export",
+		PlanID:        "11467",
+		CloudProvider: models.CloudProviderAWS,
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: this field can be set automatically by the NewAWSCloudExport function

File: examples/cloud_export_example_test.go [hunk 5]
```diff
@@ -1,85 +1,253 @@
...
+		UseBGPDeviceID: "dummy-device-id",
+		DeviceBGPType:  "dummy-device-bgp-type",
+	}
+
+	ce, err = client.CloudExports.Create(ctx, ce)
+	if err != nil {
+		return fmt.Errorf("client.CloudExports.Create: %w", err)
+	}
+
+	fmt.Printf("Created AWS cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 6]
```diff
@@ -1,85 +1,253 @@
...
+
+	fmt.Println("Getting AWS cloud export")
+	ce, err = client.CloudExports.Get(ctx, ce.ID)
+	if err != nil {
+		return fmt.Errorf("client.CloudExports.Get: %w", err)
+	}
+
+	fmt.Printf("Got AWS cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 7]
```diff
@@ -1,85 +1,253 @@
...
+	ce.Name = "updated-example-aws-export"
+	ce.Description = "Updated description"
+	ce.AWSProperties.Bucket = "updated-bucket"
+	ce.BGP.UseBGPDeviceID = "updated-bgp-device-id"
+	ce, err = client.CloudExports.Update(ctx, ce)
+	if err != nil {
+		return fmt.Errorf("client.CloudExports.Update: %w", err)
+	}
+
+	fmt.Printf("Updated cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 8]
```diff
@@ -1,85 +1,253 @@
...
-				},
-			},
+		return err
+	}
+
+	fmt.Println("Creating Azure cloud export")
+	ce := models.NewAzureCloudExport(models.CloudExportAzureRequiredFields{
+		Name:          "example-azure-export",
+		PlanID:        "11467",
+		CloudProvider: models.CloudProviderAzure,
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: can be set in constructor

File: examples/cloud_export_example_test.go [hunk 9]
```diff
@@ -1,85 +1,253 @@
...
+	}
+	ce, err = client.CloudExports.Create(ctx, ce)
 	if err != nil {
-		return fmt.Errorf("client.CloudExportAdmin.CreateCloudExport: %w", err)
+		return fmt.Errorf("client.CloudExports.Create: %w", err)
 	}
 
-	PrettyPrint(createResp.GetExport())
-	fmt.Println()
+	fmt.Printf("Created Azure cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 10]
```diff
@@ -1,85 +1,253 @@
...
-	export := createResp.GetExport()
-	export.Description = "Updated description"
-	updateResp, err := client.CloudExportAdmin.UpdateCloudExport(ctx, &cloudexportpb.UpdateCloudExportRequest{
-		Export: export,
+	fmt.Println("Creating GCE cloud export")
+	fmt.Println("Creating Azure cloud export")
+	ce := models.NewGCECloudExport(models.CloudExportGCERequiredFields{
+		Name:          "example-gce-export",
+		PlanID:        "11467",
+		CloudProvider: models.CloudProviderGCE,
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: set in constructor

File: examples/cloud_export_example_test.go [hunk 11]
```diff
@@ -1,85 +1,253 @@
...
+
+	ce, err = client.CloudExports.Create(ctx, ce)
 	if err != nil {
-		return fmt.Errorf("client.CloudExportAdmin.UpdateCloudExport: %w", err)
+		return fmt.Errorf("client.CloudExports.Create: %w", err)
 	}
 
-	PrettyPrint(updateResp.GetExport())
-	fmt.Println()
+	fmt.Printf("Created GCE cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 12]
```diff
@@ -1,85 +1,253 @@
...
+	client, err := NewClient()
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Creating IBM cloud export")
+	ce := models.NewIBMCloudExport(models.CloudExportIBMRequiredFields{
+		Name:          "example-ibm-export",
+		PlanID:        "11467",
+		CloudProvider: models.CloudProviderIBM,
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: set in constructor

File: examples/cloud_export_example_test.go [hunk 13]
```diff
@@ -1,85 +1,253 @@
...
+		UseBGPDeviceID: "dummy-device-id",
+		DeviceBGPType:  "dummy-device-bgp-type",
+	}
+
+	ce, err = client.CloudExports.Create(ctx, ce)
+	if err != nil {
+		return fmt.Errorf("client.CloudExports.Create: %w", err)
+	}
+
+	fmt.Printf("Created IBM cloud export: %+v\n", ce)
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go]: PrettyPrint

File: examples/cloud_export_example_test.go [hunk 14]
```diff
@@ -1,85 +1,253 @@
...
+func TestDemonstrateCloudExportAPIWithAWS(t *testing.T) {
 	t.Parallel()
-	err := demonstrateCloudExportAPI()
+	err := demonstrateCloudExportAPIWithAWS()
 	assert.NoError(t, err)
 }
 
-func demonstrateCloudExportAPI() error {
+func TestDemonstrateCloudExportAPIWithAzure(t *testing.T) {
+	t.Parallel()
```
Comment (mateuszmidor) [inline examples/cloud_export_example_test.go:24]: my observation is this parallel execution makes analysis of the examples output harder when running all the examples

File: kentikapi/cloud_exports_integration_test.go [hunk 15]
```diff
@@ -0,0 +1,1028 @@
...
+			FlowFound:            &wrapperspb.BoolValue{Value: false},
+			ApiAccess:            &wrapperspb.BoolValue{Value: true},
+			StorageAccountAccess: &wrapperspb.BoolValue{Value: false},
+		},
+	}
+}
+
+func codePtr(c codes.Code) *codes.Code {
+	return &c
+}
```
Comment (mateuszmidor) [inline kentikapi/cloud_exports_integration_test.go:1027]: Given that both CloudExport grpc client and server are auto-generated from the same grpc protocol files, I'm not quite sure if such integration testing presents much profit; Im afraid the manually written integration test is more prone to errors than the auto-generated client-server code under test. 
The part that might deserve testing is data transformation between models <-> pb and this can be done as unit tests. Although not that much necessary; we are simply repacking data from one GO structure to another, not dealing with any savage untyped json data. What do you think?

File: kentikapi/models/cloud_export.go [hunk 16]
```diff
@@ -0,0 +1,243 @@
+package models
+
+// NewAWSCloudExport creates a new AWS CloudExport with all required fields set.
+func NewAWSCloudExport(obj CloudExportAWSRequiredFields) *CloudExport {
+	return &CloudExport{
+		Name:          obj.Name,
+		PlanID:        obj.PlanID,
+		CloudProvider: obj.CloudProvider,
```
Comment (mateuszmidor) [inline kentikapi/models/cloud_export.go]: CloudProvider should be fixed to "aws"

File: kentikapi/models/cloud_export.go [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+		},
+	}
+}
+
+// NewAzureCloudExport creates a new AZURE CloudExport with all required fields set.
+func NewAzureCloudExport(obj CloudExportAzureRequiredFields) *CloudExport {
+	return &CloudExport{
+		Name:          obj.Name,
+		PlanID:        obj.PlanID,
+		CloudProvider: obj.CloudProvider,
```
Comment (mateuszmidor) [inline kentikapi/models/cloud_export.go]: CloudProvider should be fixed to "azure"

File: kentikapi/models/cloud_export.go [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+		},
+	}
+}
+
+// NewGCECloudExport creates a new GCE CloudExport with all required fields set.
+func NewGCECloudExport(obj CloudExportGCERequiredFields) *CloudExport {
+	return &CloudExport{
+		Name:          obj.Name,
+		PlanID:        obj.PlanID,
+		CloudProvider: obj.CloudProvider,
```
Comment (mateuszmidor) [inline kentikapi/models/cloud_export.go]: CloudProvider should be fixed to "gce"

File: kentikapi/models/cloud_export.go [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+		},
+	}
+}
+
+// NewIBMCloudExport creates a new IBM CloudExport with all required fields set.
+func NewIBMCloudExport(obj CloudExportIBMRequiredFields) *CloudExport {
+	return &CloudExport{
+		Name:          obj.Name,
+		PlanID:        obj.PlanID,
+		CloudProvider: obj.CloudProvider,
```
Comment (mateuszmidor) [inline kentikapi/models/cloud_export.go]: CloudProvider should be fixed to "ibm"

File: kentikapi/models/cloud_export.go [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+// CloudExport defines a cloud export task.
+type CloudExport struct {
+	// Read-write properties
+
+	// Type of export task.
+	Type CloudExportType
+	// Enabled specifies whether this task is enabled and intended to run or disabled.
+	Enabled *bool
```
Comment (mateuszmidor) [inline kentikapi/models/cloud_export.go:70]: why is this a pointer?

**Review by mateuszmidor — COMMENTED** [review]
New interface looks friendly.
Left a few comments.

File: examples/cloud_export_example_test.go (hunk 2 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 5 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 6 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 7 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 9 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 11 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 13 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 14 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go:24]: Yes, that's why it is disabled in instructions in README (`-parallel 1`)

Comment (danielfurman) [reply examples/cloud_export_example_test.go:24]: For me parallel examples are useful

File: kentikapi/cloud_exports_integration_test.go (hunk 15 above)
Comment (danielfurman) [reply kentikapi/cloud_exports_integration_test.go:1027]: I have to end for today, but in short: this integration test tests also conversions from models to pb and back. I could test that also via unit test. It wouldn't be easier for me and it wouldn't be noticably more performant.

If I have to choose between integration test and unit test, I pick integration test, unless it is noticably slower/harder to write. Because it tests more interactions by definition.

BTW, those tests are useful. They caught at least one nil pointer dereference in my implementation of conversions

File: kentikapi/models/cloud_export.go (hunk 20 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:70]: In short: for consistency with how other booleans are handled.

File: kentikapi/cloud_exports_integration_test.go (hunk 15 above)
Comment (mateuszmidor) [reply kentikapi/cloud_exports_integration_test.go:1027]: OK, it's already in place anyway, I just think it's an overkill

File: kentikapi/models/cloud_export.go (hunk 20 above)
Comment (mateuszmidor) [reply kentikapi/models/cloud_export.go:70]: OK

File: examples/cloud_export_example_test.go (hunk 14 above)
Comment (mateuszmidor) [reply examples/cloud_export_example_test.go:24]: OK

File: examples/cloud_export_example_test.go (hunk 4 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 8 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 10 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: examples/cloud_export_example_test.go (hunk 12 above)
Comment (danielfurman) [reply examples/cloud_export_example_test.go]: Done

File: kentikapi/cloud_exports_integration_test.go (hunk 15 above)
Comment (danielfurman) [reply kentikapi/cloud_exports_integration_test.go:1027]: It might be an overkill. I frankly don't know if UTs would be easier to write for me.

File: kentikapi/models/cloud_export.go (hunk 16 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go]: Done

File: kentikapi/models/cloud_export.go (hunk 17 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go]: Done

File: kentikapi/models/cloud_export.go (hunk 18 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go]: Done

File: kentikapi/models/cloud_export.go (hunk 19 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go]: Done

**Comment by danielfurman** [comment]
@mateuszmidor @jdziurzynski Review remarks applied. Please re-review

**Review by mateuszmidor — APPROVED** [review]

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+		CloudExports:             cloudExportsFromPayload(r.Exports),
+		InvalidCloudExportsCount: r.InvalidExportsCount,
+	}
+}
+
+func cloudExportsFromPayload(exports []*cloudexportpb.CloudExport) []models.CloudExport {
+	var result []models.CloudExport
+	for _, e := range exports {
+		ce := CloudExportFromPayload(e)
+		if ce != nil {
```
Comment (mmac-m3a) [inline kentikapi/internal/api_payloads/cloud_export.go:29]: AFAICT, `ce` should never be `nil`, unless there is a `nil` object in the list returned by the API client. Can that ever happen? If not, we should abort and report error if this test fails. Ignoring the problem is not acceptable.

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 22]
```diff
@@ -0,0 +1,232 @@
...
+			result = append(result, *ce)
+		}
+	}
+	return result
+}
+
+// CloudExportFromPayload converts cloud export payload to model.
+func CloudExportFromPayload(ce *cloudexportpb.CloudExport) *models.CloudExport {
+	if ce == nil {
+		return nil
```
Comment (mmac-m3a) [inline kentikapi/internal/api_payloads/cloud_export.go:39]: We should log at least a debug message here to make troubleshooting of the unlikely case easier.

File: kentikapi/internal/api_payloads/cloud_export.go [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+	return &models.BGPProperties{
+		ApplyBGP:       pointer.ToBool(bgp.GetApplyBgp()),
+		UseBGPDeviceID: bgp.GetUseBgpDeviceId(),
+		DeviceBGPType:  bgp.GetDeviceBgpType(),
+	}
+}
+
+func currentStatusFromPayload(cs *cloudexportpb.Status) *models.CloudExportStatus {
+	if cs == nil {
+		return nil
```
Comment (mmac-m3a) [inline kentikapi/internal/api_payloads/cloud_export.go:123]: Should log here too, because it should never happen.

File: kentikapi/models/cloud_export.go [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+// GetAllCloudExportsResponse model.
+type GetAllCloudExportsResponse struct {
+	// CloudExports holds all cloud export tasks.
+	CloudExports []CloudExport
+	// InvalidCloudExportsCount is a number of invalid cloud export tasks.
+	InvalidCloudExportsCount uint32
+}
+
+// CloudExport defines a cloud export task.
+type CloudExport struct {
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:64]: This looks pretty much a copy'n'paste of https://github.com/kentik/api-schema-public/blob/acc22cb40b784631c9ef9d9be7594228f02a3487/gen/go/kentik/cloud_export/v202101beta1/cloud_export.pb.go#L88 (including comments).
It is OK as a start, but we will need to refactor it. This model of a cloud export is IMHO unnecessarily complicated and we will likely improve it in the GA version of the cloud export API. See more comments bellow.

File: kentikapi/models/cloud_export.go [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+	APIRoot string
+	// FlowDestination is a hostname of the Kentik deployment where the data generated by this export should go,
+	// e.g. "flow.kentik.com".
+	FlowDestination string
+	// PlanID is the identifier of the Kentik plan associated with this export.
+	PlanID string
+	// CloudProvider is the cloud provider targeted by this export, e.g. AWS, Azure, GCE, IBM.
+	CloudProvider CloudProvider
+	// AWSProperties are specific to Amazon Web Services VPC flow logs exports.
+	AWSProperties *AWSProperties
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:86]: Can we turn this into a union (if golang can support something like it)? It is guaranteed that always only one set of properties (specific to the provider) is used.

File: kentikapi/models/cloud_export.go [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+	Type CloudExportType
+	// Enabled specifies whether this task is enabled and intended to run or disabled.
+	Enabled *bool
+	// Name is a short name for this export task.
+	Name string
+	// Description is an optional, longer description of this export task.
+	Description string
+	// APIRoot is a hostname of the Kentik deployment where the API calls related to this export should go,
+	// e.g. "api.kentik.com".
+	APIRoot string
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:77]: `APIRoot` and `FlowDestination` should not be exposed. We will likely remove those fields in the GA version of cloud export API.

File: kentikapi/models/cloud_export.go [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+	// Bucket is source S3 bucket to fetch VPC flow logs from.
+	Bucket string
+	// IAMRoleARN is an ARN for the IAM role to assume when fetching data or making AWS calls for this export.
+	IAMRoleARN string
+	// Region is an AWS region where the source S3 bucket resides.
+	Region string
+	// DeleteAfterRead enables the deletion of VPC flow log chunks from S3 after they have been read.
+	DeleteAfterRead *bool
+	// MultipleBuckets enables using multiple source S3 buckets.
+	MultipleBuckets *bool
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:115]: This will also be changed (or possibly removed), because it actually does not mean what the name of the field implies.

File: kentikapi/models/cloud_export.go [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+	StorageAccount string
+	// SubscriptionID is the Azure subscription ID.
+	SubscriptionID string
+	// SecurityPrincipalEnabled enables security principal.
+	SecurityPrincipalEnabled *bool
+}
+
+// GCEProperties are specific to Google Cloud export.
+type GCEProperties struct {
+	// Project is a GCE project name.
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:134]: I'd bet that GCE will need region or location at some point too.

File: kentikapi/models/cloud_export.go [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+	Project string
+	// Subscription is a GCE subscription name.
+	Subscription string
+}
+
+// IBMProperties are specific to IBM Cloud exports.
+type IBMProperties struct {
+	// Bucket is an IBM bucket.
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:142]: ... and IBM too.

File: kentikapi/models/cloud_export.go [hunk 30]
```diff
@@ -0,0 +1,239 @@
...
+// CloudExportAWSRequiredFields is subset of fields required to create an AWS CloudExport.
+type CloudExportAWSRequiredFields struct {
+	Name          string
+	PlanID        string
+	AWSProperties AWSPropertiesRequiredFields
+}
+
+// AWSPropertiesRequiredFields is subset of AWSProperties required to create an AWS CloudExport.
+type AWSPropertiesRequiredFields struct {
+	Bucket string
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:175]: I think that `region` will become required soon.

File: kentikapi/models/cloud_export.go [hunk 31]
```diff
@@ -0,0 +1,239 @@
...
+	ErrorMessage string
+	// FlowFound informs whether flow logs were found.
+	FlowFound            *bool
+	APIAccess            *bool
+	StorageAccountAccess *bool
+}
+
+// CloudExportAWSRequiredFields is subset of fields required to create an AWS CloudExport.
+type CloudExportAWSRequiredFields struct {
+	Name          string
```
Comment (mmac-m3a) [inline kentikapi/models/cloud_export.go:168]: Why not use a common struct for common required attributes and rely on [promoted struct fields](https://golangdocs.com/structs-in-golang)? Something like this:

```go
type CommonRequiredFields struct {
    Name    string
    PlanID  string
}

type CloudExportAWSRequiredFields struct {
    CommonRequiredFields
    AWSProperties AWSPropertiesRequiredFields
}
```

**Review by mmac-m3a — APPROVED** [review]
Nice! I'll merge the code, but please consider my comments for future iterations.

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 21 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:29]: I was not able to hit that corner case with integration test.
Anyway, I'll treat nil exports in response payload as invalid and return error to the user.

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 22 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:39]: I'll treat nil exports in response payload as invalid and return error to the user. So the log will not be necessary

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 23 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:123]: I'll log a warning here.

File: kentikapi/models/cloud_export.go (hunk 25 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:86]: There are no union types in Go, but they can be emulated with interfaces. Go proto generator uses that technique: https://github.com/kentik/api-schema-public/blob/master/gen/go/kentik/cloud_export/v202101beta1/cloud_export.pb.go#L118
It's described in http://www.jerf.org/iri/post/2917
I found this approach harder to use than using multiple ordinary fields. For example modifying one field of a property is easier that way: `ce.AWSProperties.Bucket = "updated-bucket"`
(https://github.com/kentik/community_sdk_golang/blob/839c0e7fade7e7d08693d5f23c1ef3eadceb686f/examples/cloud_export_example_test.go#L101).

When using interfaces we could simplify such case by adding getters for properties. We would have: `ce.GetAWSProperties().Bucket = "updated-bucket"`

I'll implement that on Monday and see how it behaves.

File: kentikapi/models/cloud_export.go (hunk 26 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:77]: I'll remove those fields

File: kentikapi/models/cloud_export.go (hunk 27 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:115]: Should I remove it now?

File: kentikapi/models/cloud_export.go (hunk 28 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:134]: Should I add such fields now? YAGNI tells me no

File: kentikapi/models/cloud_export.go (hunk 29 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:142]: Should I add such fields now? YAGNI tells me no

File: kentikapi/models/cloud_export.go (hunk 30 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:175]: Should I add such fields now? YAGNI tells me no

File: kentikapi/models/cloud_export.go (hunk 31 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:168]: 1. I initially tried that, but I found usage to be too verbose and changed it.:

```go
ce := models.NewAzureCloudExport(models.CloudExportAzureRequiredFields{
		CommonCloudExportRequiredFields: models.CommonCloudExportRequiredFields{
			Name:   "example-azure-export",
			PlanID: "11467",
		},
		AzureProperties: models.AzurePropertiesRequiredFields{
			Location:       "dummy-location",
			ResourceGroup:  "dummy-rg",
			StorageAccount: "dummy-sa",
			SubscriptionID: "dummy-sid",
		},
	})
```

2. Other option is to omit struct key names, but it is generally discouraged:

```go
ce := models.NewAzureCloudExport(models.CloudExportAzureRequiredFields{
		models.CommonCloudExportRequiredFields{
			Name:   "example-azure-export",
			PlanID: "11467",
		},
		models.AzurePropertiesRequiredFields{
			Location:       "dummy-location",
			ResourceGroup:  "dummy-rg",
			StorageAccount: "dummy-sa",
			SubscriptionID: "dummy-sid",
		},
	})
```

3. I prefer simpler API for cost of little harder maintainability:
  
```go
ce := models.NewAzureCloudExport(models.CloudExportAzureRequiredFields{
		Name:   "example-azure-export",
		PlanID: "11467",
		AzureProperties: models.AzurePropertiesRequiredFields{
			Location:       "dummy-location",
			ResourceGroup:  "dummy-rg",
			StorageAccount: "dummy-sa",
			SubscriptionID: "dummy-sid",
		},
	})
```

@mateuszmidor @jdziurzynski Any opinions on that?

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 21 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:29]: Addresed in #88

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 22 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:39]: Addresed in #88

File: kentikapi/internal/api_payloads/cloud_export.go (hunk 23 above)
Comment (danielfurman) [reply kentikapi/internal/api_payloads/cloud_export.go:123]: Addresed in #88

File: kentikapi/models/cloud_export.go (hunk 26 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:77]: Addresed in #88

File: kentikapi/models/cloud_export.go (hunk 25 above)
Comment (danielfurman) [reply kentikapi/models/cloud_export.go:86]: Addressed in https://github.com/kentik/community_sdk_golang/pull/89
